cmake: Allow semver style EXTRAVERSION

nrbrook opened the pull request:
Semver versions are commonly in the form X.X.X-a.1, X.X.X-beta.2, etc. However, currently EXTRAVERSION cannot include ".". This change allows it.

pdgendt commented:
Please update the commit subject as described in the [commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) and add a `Signed-off-by` to the commit.

nrbrook committed:
cmake: Allow semver style EXTRAVERSION

cmake: Allow semver style EXTRAVERSION

Semver versions are commonly in the form X.X.X-a.1, X.X.X-beta.2, etc.
However, currently EXTRAVERSION cannot include ".". This change allows it.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook commented:
I think I have done it correctly now, please check and advise if not.

A couple things I was unsure about:

1. The documentation for EXTRAVERSION should be updated, I couldn't see it in the repo.
2. Because EXTRAVERSION is not used in the numerical representation `APPVERSION`, will bumping the beta version in EXTRAVERSION be recognized as a distinct version when performing DFU? I'm not sure how APPVERSION is used.

tejlmand commented on cmake/modules/version.cmake:
```diff
@@ -63,7 +63,7 @@ foreach(type file IN ZIP_LISTS VERSION_TYPE VERSION_FILE)
   set(${type}_VERSION_TWEAK ${CMAKE_MATCH_1})
 
-  string(REGEX MATCH "EXTRAVERSION = ([a-z0-9]*)" _ ${ver})
+  string(REGEX MATCH "EXTRAVERSION = ([a-z0-9\.]*)" _ ${ver})
```
should we consider to allow a hyphen `-` as well.
That's is valid in pre-release version.
https://semver.org/#spec-item-9
> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-].
> ...
> Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92, **1.0.0-x-y-z.--**.

Check also: https://regex101.com/r/vkijKf/1/

nrbrook replied:
I wasn't aware of that. I can add it to allow that format?

nordicjm replied:
If you feel like updating PR with it

tejlmand replied:
I think it would be nice, else we probably just end back later with someone else missing the hyphen `-` in the format.

nrbrook replied:
I've added it.

nrbrook committed:
cmake: Added "-" support to EXTRAVERSION

cmake: Added "-" support to EXTRAVERSION.
Added to the regex and update the documentation.
Also updated the examples in the documentation.

Signed-off-by: Nick Brook <nrbrook@gmail.com>

nrbrook commented:
> 2. Because EXTRAVERSION is not used in the numerical representation `APPVERSION`, will bumping the beta version in EXTRAVERSION be recognized as a distinct version when performing DFU? I'm not sure how APPVERSION is used.

To answer my own question, EXTRAVERSION is not used during DFU, so to bump the version in a way that will ensure the update is performed, TWEAK must also be modified.
